pkg/gadgets/network-graph/tracer/tracer.go

```diff
@@ -0,0 +1,367 @@
+}
+
+func protoString(proto int) string {
+	// proto definitions:
+	// https://www.iana.org/assignments/protocol-numbers/protocol-numbers.xhtml
+	protoStr := fmt.Sprintf("UNKNOWN#%d", proto)
+	switch proto {
+	case 6:
+		protoStr = "tcp"
+	case 17:
+		protoStr = "udp"
+	}
+	return protoStr
+}
+
+func (t *Tracer) containerQuarkToKey(quark uint64) string {
+	key := "NotFound"
+	for k, v := range t.attachments {
+		if quark == v.containerQuark {
+			key = k
+			break
+		}
+	}
+	return key
+}
+
+func (t *Tracer) popIterator() ([]Edge, error) {
+	file, err := t.dumpIter.Open()
+	if err != nil {
+		return nil, fmt.Errorf("cannot open iter instance: %w", err)
+	}
+	defer file.Close()
+
+	edges := []Edge{}
+
+	scanner := bufio.NewScanner(file)
+	for scanner.Scan() {
+		var containerQuark uint64
+		var pktType, proto, port int
+		var ipStr string
+
+		text := scanner.Text()
+		matchedElems, err := fmt.Sscanf(text, "%d %d %d %d %s",
+			&containerQuark,
+			&pktType,
+			&proto,
+			&port,
+			&ipStr)
+		if err != nil {
+			return nil, fmt.Errorf("failed to parse %q: %w", text, err)
+		}
+		if matchedElems != 5 {
+			return nil, fmt.Errorf("failed to parse bpf iterator, expected 5 matched elements had %d", matchedElems)
+		}
+		ip := net.ParseIP(ipStr)
+
+		edges = append(edges, Edge{
+			Key:     t.containerQuarkToKey(containerQuark),
+			PktType: pktTypeString(pktType),
+			IP:      ip,
+			Proto:   protoString(proto),
+			Port:    port,
+		})
+	}
+
+	return edges, nil
+}
+
+// popFallback extracts data from the eBPF map but without the BPF Iterator
+func (t *Tracer) popFallback() ([]Edge, error) {
+	edges := []Edge{}
+	var prev *C.struct_graph_key_t = nil
+	key := C.struct_graph_key_t{}
+	err := t.graphMap.NextKey(nil, unsafe.Pointer(&key))
+	if err != nil {
+		return []Edge{}, nil
+	}
+
+	for {
+		ip := make(net.IP, 4)
+		binary.BigEndian.PutUint32(ip, uint32(C.htonl(key.ip)))
+		edges = append(edges, Edge{
+			Key:     t.containerQuarkToKey(uint64(key.container_quark)),
+			PktType: pktTypeString(int(key.pkt_type)),
+			IP:      ip,
+			Proto:   protoString(int(key.proto)),
+			Port:    int(C.htons(key.port)),
+		})
+
+		if err := t.graphMap.Delete(key); err != nil {
+			return nil, fmt.Errorf("error deleting key: %w", err)
+		}
+
+		prev = &key
+		if err := t.graphMap.NextKey(unsafe.Pointer(prev), unsafe.Pointer(&key)); err != nil {
+			if errors.Is(err, ebpf.ErrKeyNotExist) {
+				break
+			}
+			return nil, fmt.Errorf("error getting next key: %w", err)
+		}
+	}
+	return edges, nil
+}
+
+func (t *Tracer) releaseLink(key string, l *link) {
+	unix.Close(l.sockFd)
+	l.collectionProg.Close()
+	delete(t.attachments, key)
+}
+
+func (t *Tracer) Detach(key string) error {
+	if l, ok := t.attachments[key]; ok {
+		l.users--
+		if l.users == 0 {
+			t.releaseLink(key, l)
+		}
+		return nil
+	} else {
+		return fmt.Errorf("key not attached: %q", key)
+	}
+}
+
+func (t *Tracer) Close() {
+	for key, l := range t.attachments {
+		t.releaseLink(key, l)
+	}
```

Comment:
missing `t.collectionMap.Close()` and `t.collectionIter.Close()`?

Reply:
Fixed with the various `*Objects.Close()`